### PR TITLE
Add native python functional model

### PR DESCRIPTION
**Pull request by Kuree**
## Usage example:
DUT:
```Python
class dut(m.Circuit):
    io = m.IO(A=m.In(m.Bits[4]), B=m.In(m.Bits[4]), O=m.Out(m.Bits[4]))
    io.O @= io.A + io.B
```

Functional model (used for demonstrate class constructor and method call):
```Python
class Model:
    @sv(b=DataType.UByte)
    def __init__(self, b):
        self.b = b
 
   @sv(a=DataType.UByte)
    def add(self, a):
        return a + self.b
```

Test code:
```Python

a_value = 1
tester = fault.Tester(dut)
tester.poke(tester.circuit.A, a_value)

model = tester.Var("model", Model)
tester.poke(model, tester.make_call(Model, a_value))

# start testing. using a loop
loop = tester.loop(10)
loop.poke(tester.circuit.B, loop.index)
loop.eval()
loop.expect(tester.circuit.O, tester.make_call(model.add, loop.index))
```

## Nee functions/class introduced
1. Add `make_call` to tester class
2. Add `CallExpression`  to `expression`
3. Extend `Var` to support `var.foo`, where `var` is declared `tester.Var`, and `foo` is a function name. Compile time check to ensure correctness

## Code refactored
1. Refactor template the support start and exit code region. This can be used for simulator environment setup and teardown.

## Simulator tested
- Xcelium/Incisive
- VCS
- Verilator

## Discussion
This should solve #280 fundamentally by allowing arbitrary Python code execution inside a simulator. It also enables runtime model check with values unknown during compilation time.

Most of the heavy lifting is done via the `pysv` package. Currently there are some limitation on how the code should rewritten. Notably how the class and functions import are handled. If there is enough use case for that I can spend some time to solve that issue.

### Roadmap
- I will continue work on bringing the support for arrays. Ideally it should not introduce any changes to fault.
- I also talked to @sgherbst  and he suggested that it might also be possible to do something similar with SPICE simulator. I will look into that.

### CODE REVIEW

**Comment by Kuree** [comment]
Oddly the tests passed for fpga on buildkite but not the normal one. It seems like conda has some weird python system path setup. I will take a look into it. One workaround is to use `venv` instead of conda on `cadence7`, which I've successfully tested.

**Comment by leonardt** [comment]
The conda env is setup with the following commands:
```
conda create -p ./conda_env -m -y pip numpy python=3.7.3
export PATH=$PWD/conda_env/bin:$PATH
```
so it creates a local environment in the current working directory.  I'm guessing we need to configure pysv to use the correspond library load path? Is there a way we can point it to a custom directory, e.g. with an environment variable? I guess this is similar to how `venv` works with a special directory, so is there anything special done to configure that?

Could it be that `pysv` is building/linking with the wrong python? How does it find out which one to use? Since we're only editing `PATH`, maybe we're not setting the right environment variable like PYTHONHOME?

**Comment by leonardt** [comment]
Relevant part of the workflow: https://github.com/leonardt/fault/blob/master/.buildkite/pipeline.yml#L12-L13

**Comment by Kuree** [comment]
For libraries, pysv uses the `sys.path` of whatever Python it is running with and then have the compiled runtime load that, see [here](https://pysv.readthedocs.io/advanced/internal.html#foreign-definitions).

As for linking, it uses pybind11's built in capability to detect currently python executable and libraries. The error seems to look like library loading issue.

**Comment by Kuree** [comment]
@leonardt I tried to prune unnecessary imports from pysv's side to avoid switching `sys.path`, still can't reproduce the error.

Here is my working directory on cadence7: `/tmp/fault-pysv/`

```
pytest tests/test_pysv.py -v
```

Conda is created at `/tmp/fault-pysv/conda_env` using the exact same command.

Tried with several Linux distro with conda and still can't reproduce the error. Would be nice if all the tests are containerized and easy to reproduce.

File: fault/system_verilog_target.py [hunk 1]
```diff
@@ -681,6 +703,9 @@ def generate_code(self, actions, power_args):
             port_list.extend(result)
         port_list = f',\n{2*self.TAB}'.join(port_list)
 
+        if len(self.pysv_funcs) > 0:
+            self.imports.append("pysv")
```
Comment (leonardt) [inline fault/system_verilog_target.py:707]: Looking at 538, which guards against adding pysv more than once, could this add a pysv import more than once? Perhaps we could just simply make `imports` a set?

File: fault/system_verilog_target.py [hunk 2]
```diff
@@ -766,19 +803,25 @@ def run(self, actions, power_args=None):
...
-        if self.simulator == 'ncsim' or self.simulator == "incisive":
-            # Compile and run simulation
-            cmd_file = self.write_cadence_tcl()
-            sim_cmd = self.ncsim_cmd(sources=vlog_srcs, cmd_file=cmd_file)
-            sim_err_str = None
-            # Skip "bin_cmd"
-            bin_cmd = None
-            bin_err_str = None
-        elif self.simulator == 'xcelium':
+        if self.simulator == 'ncsim' or self.simulator == "incisive" or self.simulator == "xcelium":
```
Comment (leonardt) [inline fault/system_verilog_target.py:816]:
```suggestion
        if self.simulator in {"ncsim", "incisive", "xcelium"}:
```

File: fault/verilator_target.py [hunk 3]
```diff
@@ -730,6 +751,14 @@ def run(self, actions, verilator_includes=None, num_tests=0,
         else:
             env = None
 
+        # codegen the c++ binding if needed
+        if len(self.pysv_funcs) > 0:
+            header = os.path.join(self.directory, "pysv.hh")
+            pysv.generate_cxx_binding(self.pysv_funcs, filename=header)
+            lib_path = pysv.compile_lib(self.pysv_funcs, os.path.join(self.directory, "obj_dir"))
+            if env is None:
+                env = {"LD_LIBRARY_PATH": os.path.realpath(os.path.dirname(lib_path))}
```
Comment (leonardt) [inline fault/verilator_target.py:760]: Should we have an else case here where we either insert LD_LIBRARY_PATH (if it doesn't exit) or append to it (if it exists)?

**Review by leonardt — APPROVED** [review]
LGTM, finally got CI passing, some minor comments to address before merging

File: fault/system_verilog_target.py (hunk 1 above)
Comment (Kuree) [reply fault/system_verilog_target.py:707]: I was worried about non-deterministic code generation result. It seems like none of the tests are directly comparing the code output so I think using set would be better.

**Pull request merged**